### PR TITLE
APD-14038:Handle schema change

### PR DESCRIPTION
**Pull request by qilinext**
not ready for merge, create for discussion

### CODE REVIEW

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java [hunk 1]
```diff
@@ -91,7 +91,7 @@ public BigQueryWriter(int retries, long retryWaitMs) {
   protected InsertAllRequest createInsertAllRequest(PartitionedTableId tableId,
                                                     Collection<InsertAllRequest.RowToInsert> rows) {
     return InsertAllRequest.newBuilder(tableId.getFullTableId(), rows)
-        .setIgnoreUnknownValues(false)
+        .setIgnoreUnknownValues(true)
```
Comment (qilinext) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java:94]: since `fees_to_pay` is back, setting this to true can make the big query working

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java [hunk 2]
```diff
@@ -402,12 +412,17 @@ private boolean isValidFieldAddition(Field newField) {
     Map<String, Field> existingSchemaFields = schemaFields(existingSchema);
     Map<String, Field> proposedSchemaFields = schemaFields(proposedSchema);
     List<Field> newSchemaFields = new ArrayList<>();
+    for (Map.Entry<String, Field> entry : existingSchemaFields.entrySet()) {
```
Comment (qilinext) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:415]: if we really want to handle missing column, this can work

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java (hunk 1 above)
Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java:94]: we can probably set this as a parameter we can pass to big query connector

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java (hunk 2 above)
Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:415]: there is no way to handle `type change`

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java (hunk 1 above)
Comment (kppullin) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java:94]: Which error is this related to?  Unknown values on the BQ side (or in other words, this handles dropping a column)?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java [hunk 3]
```diff
@@ -402,12 +412,17 @@ private boolean isValidFieldAddition(Field newField) {
...
+      newSchemaFields.add(entry.getValue());
+    }
+
     for (Map.Entry<String, Field> entry : proposedSchemaFields.entrySet()) {
       if (!existingSchemaFields.containsKey(entry.getKey())) {
         newSchemaFields.add(entry.getValue().toBuilder().setMode(Field.Mode.NULLABLE).build());
```
Comment (kppullin) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:421]: Does the `add` here replace the existing field, if it exists?  Should we only add fields from `existingSchemaFields` that are missing from `newSchemaFields`?

Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:421]: the `newSchemaFields` in fact is the final schema, in order to deal with `deleted columns`, you need to maintain the deleted on to it, the previous logic doesn't handle deleted columns, so it throw a missing column expection

Comment (kppullin) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:421]: OK I see it now.  First we add all existing fields into new fields, and then add any actual new fields when looping through `proposedSchemaFields` (with the `if !existing.contains(key)` statement preventing duplicates).

Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:421]: before the logic is: if the proposed new fields is not exist in the existing schema, it will add the new nullable field, however it doesn't handle if proposed new fields miss column (or it is by design will throw a missing column exception)

Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:421]: now the logic is whatever fields in the existing schema, will be added to the final schema, so it can handle the `missing column` exception

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java (hunk 1 above)
Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java:94]: when table is created, `fees_to_pay` is included in the schema
because there are some data without `fees_to_pay`, so it throws `missing column` exception
`fees_to_pay`, also the counts matches

Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java:94]: let me do one more test, if revert this to false (but with missing column handling change) to see what exactly the error it will throw

Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java:94]: I tested out with this parameter on and off:

1. setIgnoreUnknownValues(false)
it fails after 10 attempts
restart --> working fine, saw values of `fees_to_pay` comes back as well
2. setIgnoreUnknownValues(true) `fees_to_pay `
it success, however `fees_to_pay` all null

The above testings are based on the new logic to `add delete column` back to the schema
if without this logic, it will fail throwing 'Field fees_to_pay is missing in new schema' error

to summarize: seems better to keep setIgnoreUnknownValues(false)

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java [hunk 4]
```diff
@@ -402,15 +402,28 @@ private boolean isValidFieldAddition(Field newField) {
...
     Map<String, Field> proposedSchemaFields = schemaFields(proposedSchema);
     List<Field> newSchemaFields = new ArrayList<>();
-    // for handle column delete situation
-    for (Map.Entry<String, Field> entry : existingSchemaFields.entrySet()) {
-      newSchemaFields.add(entry.getValue());
-    }
-    for (Map.Entry<String, Field> entry : proposedSchemaFields.entrySet()) {
-      if (!existingSchemaFields.containsKey(entry.getKey())) {
-        newSchemaFields.add(entry.getValue().toBuilder().setMode(Field.Mode.NULLABLE).build());
+    if (BigQuerySinkConfig.ALLOW_DELETE_COLUMN_CONFIG.equals("true")) {
```
Comment (kppullin) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:405]: Nice, thanks!

**Review by kppullin — APPROVED** [review]

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java (hunk 4 above)
Comment (qilinext) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: just tested, the hack way `BigQuerySinkConfig.ALLOW_DELETE_COLUMN_CONFIG` get the flag's value is always return false, then I am doing the normal way as other parameter settings, this works fine now
@kppullin
